src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenCapturing.cs

```diff
@@ -0,0 +1,465 @@
+        }
+
+        private string MakeLocalFunc(int nameIndex, string captureExpression)
+            => $@"int Local_{nameIndex}() => {captureExpression};";
+
+        private string MakeCaptureExpression(IList<int> varsToCapture, CaptureContext ctx)
+        {
+            var varNames = new List<string>();
+            for (int varDepth = 0; varDepth < varsToCapture.Count; varDepth++)
+            {
+                var variablesByScope = ctx.VariablesByScope;
+                // Do we have any variables in this scope depth?
+                // If not, initialize an empty list
+                if (variablesByScope.Count <= varDepth)
+                {
+                    variablesByScope.Add(new List<string>());
+                }
+
+                var varsAtCurrentDepth = variablesByScope[varDepth];
+
+                int numToCapture = varsToCapture[varDepth];
+                int numVarsAvailable = variablesByScope[varDepth].Count;
+                // If we have enough variables to capture in the context
+                // just add them
+                if (numVarsAvailable >= numToCapture)
+                {
+                    // Capture the last variables added since if there are more
+                    // vars in the context than the max vars to capture we'll never
+                    // have code coverage of the newest vars added to the context.
+                    varNames.AddRange(varsAtCurrentDepth
+                        .Skip(numVarsAvailable - numToCapture)
+                        .Take(numToCapture));
+                }
+                else
+                {
+                    // Not enough variables in the context -- add more
+                    for (int i = 0; i < numToCapture - numVarsAvailable; i++)
+                    {
+                        varsAtCurrentDepth.Add($"captureVar_{ctx.CaptureNameIndex++}");
+                    }
+
+                    varNames.AddRange(varsAtCurrentDepth);
+                }
+            }
+
+            return varNames.Count == 0 ? "0" : string.Join(" + ", varNames);
+        }
+
+        /// <summary>
+        /// Generates all combinations of distributing a sum to a list of subsets.
+        /// This is equivalent to the "stars and bars" combinatorics construction.
+        /// </summary>
+        private IEnumerable<IList<int>> GenerateAllSetCombinations(int sum, int numSubsets)
+        {
+            Assert.True(numSubsets > 0);
+            return GenerateAll(sum, 0, Array.Empty<int>());
+
+            IEnumerable<int[]> GenerateAll(
+                int remainingSum,
+                int setIndex, // 0-based index of subset we're generating
+                int[] setsSoFar)
+            {
+                for (int i = 0; i <= remainingSum; i++)
+                {
+                    var newArray = new int[setIndex + 1];
+                    Array.Copy(setsSoFar, newArray, setIndex);
+                    newArray[setIndex] = i;
```

Comment:
Would this be slightly simpler if the sets were `ImmutableList<int>` rather than `int[]`?